tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/modules/ServiceCommand.java

```diff
@@ -0,0 +1,294 @@
+/*******************************************************************************
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ *
+ *******************************************************************************/
+
+package com.liferay.ide.project.core.modules;
+
+import java.io.InputStream;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+
+import aQute.remote.api.Agent;
+import aQute.remote.api.Event;
+import aQute.remote.api.Supervisor;
+import aQute.remote.util.AgentSupervisor;
+
+/**
+ * @author Lovett Li
+ */
+public class ServiceCommand {
+
+    private String _serviceName;
+
+    public ServiceCommand()
+    {
+    }
+    public ServiceCommand( String serviceName )
+    {
+        _serviceName = serviceName;
+    }
+
+    public String[] execute() throws Exception
+    {
+        ServicesSupervisor supervisor = null;
+        String[] result = null;
+
+        synchronized( ServiceCommand.class )
+        {
+
+            try
+            {
+                supervisor = connectRemote();
+
+                if( _serviceName == null )
+                {
+                    result = getServices( supervisor );
+                }
+                else
+                {
+                    result = getServiceBundle( _serviceName, supervisor );
+                }
+
+                return result;
+            }
+            finally
+            {
+                if( supervisor != null )
+                {
+                    supervisor.close();
+                }
+
+            }
+        }
+    }
+
+    private ServicesSupervisor connectRemote() throws Exception
+    {
+        if( !canConnect( "localhost", Agent.DEFAULT_PORT ) )
+        {
+            return null;
+        }
+
+        ServicesSupervisor supervisor = new ServicesSupervisor();
+        supervisor.connect( "localhost", Agent.DEFAULT_PORT );
+
+        if( !supervisor.getAgent().redirect( -1 ) )
+        {
+            supervisor.close();
+            return null;
+        }
+
+        return supervisor;
+    }
+
+    static boolean canConnect( String host, int port )
```

Comment:
THere should be a "canConnect" in other parts of our codebase in some Util.  Call that one instead and don't duplicate the code here.